### ISSUE TITLE
Global DB failure PouchDB/IndexedDB caused by different ID

### ISSUE THREAD

**Issue by bogdanlupuro** [pouchdb/pouchdb#7020]
### Issue
We encountered on a user database a global database failure making any read/write of data impossible.  No syncing is possible or allDocs or anything, It goes to the idb error handler and closes the connection. No way of recovery from what I've seen.

It tries to get a property '_attachments' from a doc to sync it to the server. But when it tries to get it, the document is undefined and results in a error and consequently in a global DB error and close. When trying to do an allDocs also results in a error. 

Affected code area is:
```javascript
function fetchDocAsynchronously(metadata, row, winningRev$$1) {
    var key = metadata.id + "::" + winningRev$$1;
    docIdRevIndex.get(key).onsuccess = function onGetDoc(e) {
        row.doc = decodeDoc(e.target.result);
        if (opts.conflicts) {
            var conflicts = collectConflicts(metadata);
            if (conflicts.length) {
                row.doc._conflicts = conflicts;
            }
        }
        fetchAttachmentsIfNecessary(row.doc, opts, txn);
    };
}
function fetchAttachmentsIfNecessary(doc, opts, txn, cb) {
    var attachments = Object.keys(doc._attachments || {});
    if (!attachments.length) {
        return cb && cb();
    }
....
```

doc is undefined, so doc._attachments gives the above error. On close inspection this happens because a document stored in the DB: lm_note_3b548873-bc4e-4b4b-240b-7042c059a4bd::1-83e5ef9a59e34693a2fb8faf2f27f75b When it tries and get the document id does not find also the containing document so it only returns:

```json
{
  "id": "lm_note_3b548873-bc4e-4b4b-240b-7042c059a4bd",
  "key": "lm_note_3b548873-bc4e-4b4b-240b-7042c059a4bd",
  "value": {
    "rev": "1-83e5ef9a59e34693a2fb8faf2f27f75b"
  },
  "doc": undefined
}
```

On a even closer inspection I noticed that in the 3 different places in which the document is stored in IndexededDB by PouchDB it has a slighty different ID. Check pictures:
lm_note_3b548873-**bc4e**-4b4b-240b-7042c059a4bd - deletedOrLocal
lm_note_3b548873-**bc4**-4b4b-240b-7042c059a4bd::1-83e5ef9a59e34693a2fb8faf2f27f75b - _doc_id_rev
lm_note_3b548873-**bc4e**-4b4b-240b-7042c059a4bd::1-83e5ef9a59e34693a2fb8faf2f27f75b - by-sequence

Now this different ID does not make any sense and I can almost guarantee that no manual alteration was done.

### Info
- Environment: Browser/hybrid app
- Platform: Chrome
- Adapter: IndexedDB
- Server: only local db

### Reproduce
Can you think of reasons why the ID changed, missing an "e" and adding a space at the end, and also how we can have fail-safes and recovery options, as to not have a global failure.

I can reproduce the error all the time which a cache of the local DB, but I cannot make a example because I cannot manage to write the different ID in different places. 
Are there any ID manipulations?

Id first place
![doc1](https://user-images.githubusercontent.com/16715998/35052763-afd9bbfa-fba8-11e7-88cb-7913289f997f.PNG)
Id second place
![doc2](https://user-images.githubusercontent.com/16715998/35052765-aff15030-fba8-11e7-8f21-ffa3f67c102c.PNG)
Id third place
![doc3](https://user-images.githubusercontent.com/16715998/35052766-b00de3f8-fba8-11e7-90e8-7cf59ae8bf43.PNG)
Error
![alldocserror](https://user-images.githubusercontent.com/16715998/35052767-b026f884-fba8-11e7-9710-59095a52912a.PNG)

**Comment by daleharvey** [comment]
There is nowhere we manipulate ids so very confused at how that could possibly happen, have never seen it reported either, would be interested to know however nowhere I could point to.

As for it causing a global failure, be happy to merge ways to mitigate this, usually a txn.onabort or a try catch will prefent a global failure

**Comment by bogdanlupuro** [comment]
Thanks for the reply, indeed it is very strange and I cannot figure either
how this could happen.

I would be happy to provide a PR to catch this error. Would you have any
idea on how to unit test this, so that I get a response with a undefined
document? I think that should not be possible in PouchDB without a
corrupted DB.

Thanks,
Bogdan

**Comment by daleharvey** [comment]
https://github.com/pouchdb/pouchdb/blob/master/tests/component/test.read_only_replication.js is an example of a test where you can intercept and mock results from the database, there is also https://github.com/pouchdb/pouchdb/blob/master/tests/component/test.read_only_replication.js if its not enough to do that at the server response level

**Comment by bogdanlupuro** [comment]
I have done some more careful reading and after figuring out how pouchdb and idb work I noticed that this might be a problem from IDB somehow. The problem is the the _doc_id_rev index is not constructed correctly. We can see clearly the the KEY is not correct - it is different from the _doc_id_rev field from the document. How that get's corrupted is extremely weird as it slightly changed. 

There is no way to reproduce this as it is only how IDB automatically creates the index. 

I created a PR #7051 which works around this problem and at least if this would happen for a random document, at least everything else would work fine, and will not result in global failure.
